src/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -125,30 +127,104 @@ public HttpClient(HttpMessageHandler handler, bool disposeHandler)
-                r.EnsureSuccessStatusCode();
-                HttpContent c = r.Content;
-                return c != null ? c.ReadBufferedContentAsByteArray() : Array.Empty<byte>();
-            }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.DenyChildAttach, TaskScheduler.Default);
+                // Make sure it completed successfully.
+                responseMessage.EnsureSuccessStatusCode();
+                
+                // Get the response content.
+                HttpContent c = responseMessage.Content;
+                if (c != null)
+                {
+                    HttpContentHeaders headers = c.Headers;
+                    using (Stream responseStream = await c.ReadAsStreamAsync().ConfigureAwait(false))
+                    {
+                        long? contentLength = headers.ContentLength;
+                        Stream buffer; // declared here to share the state machine field across both if/else branches
+
+                        if (contentLength.HasValue)
+                        {
+                            // If we got a content length, then we assume that it's correct and create a MemoryStream
+                            // to which the content will be transferred.  That way, assuming we actually get the exact
+                            // amount we were expecting, we can simply return the MemoryStream's underlying buffer.
+                            buffer = new HttpContent.LimitMemoryStream(_maxResponseContentBufferSize, (int)contentLength.GetValueOrDefault());
+                            await responseStream.CopyToAsync(buffer).ConfigureAwait(false);
+                            if (buffer.Length > 0)
+                            {
+                                return ((HttpContent.LimitMemoryStream)buffer).GetSizedBuffer();
+                            }
+                        }
+                        else
+                        {
+                            // If we didn't get a content length, then we assume we're going to have to grow
+                            // the buffer potentially several times and that it's unlikely the underlying buffer
+                            // at the end will be the exact size needed, in which case it's more beneficial to use
+                            // ArrayPool buffers and copy out to a new array at the end.
+                            buffer = new HttpContent.LimitArrayPoolWriteStream(_maxResponseContentBufferSize);
+                            try
+                            {
+                                await responseStream.CopyToAsync(buffer).ConfigureAwait(false);
+                                if (buffer.Length > 0)
+                                {
+                                    return ((HttpContent.LimitArrayPoolWriteStream)buffer).ToArray();
+                                }
+                            }
+                            finally { buffer.Dispose(); }
```

Comment:
```using``` uses a temporary for the target of the using if the target isn't declared in the using, which means we end up getting another field in the state machine.  Ideally I would have just used a using with the variable defined in it, but that then leads to two fields as well, one for this one and one for the else case, due to https://github.com/dotnet/roslyn/issues/15290.